File: Matlab/P300speller_predict_letter.m
% P300Speller Letter Classification Training ver1: SWLDA
%
% Kyungho Won
%
% [ reference ]
% data: BCI2000, 32 Biosemi2, 55 subjects
% target text: BRAIN, POWER / SUBJECT, NEURONS, IMAGINE, QUALITY
% target : standard = 30 : 150
%
% [ Stage ]
%  1. Pre-processing: bandpass filtering, extracting triggers
%     - freq = [0.5 10]
%     - frame = [0 600]
%     - baseline = [-200 0]
%
%  2. Apply SWLDA: getting w and w0 from SWLDA
%     - y = w0 + w' * feature
%     - class = sign(y)
%  * In code,
%  predictor = sign(feature(:, INMODEL) * W(INMODEL) + STATS.intercept)
%   INMODEL (selected features)
%   W (selected weights)
%   STATS.intercept: W0
clear; clc;

ch = 1:32; % select channels
% select_ch = find(ismember({biosemi32_locs.labels}, ...
%     {'FZ', 'Cz', 'Pz', 'CP1', 'CP2'}));
spellermatrix = ...
    ['A', 'B', 'C', 'D', 'E', 'F', ...
    'G', 'H', 'I', 'J', 'K', 'L', ...
    'M', 'N', 'O', 'P', 'Q', 'R', ...
    'S', 'T', 'U', 'V', 'W', 'X', ...
    'Y', 'Z', '1', '2', '3', '4', ...
    '5', '6', '7', '8', '9', '_'];
Configs_P3speller = struct('seq_code', 1:12, 'full_repeat', 15, ...
    'spellermatrix', spellermatrix);
Params_P3speller = struct('freq', [0.5 10], 'frame', [0 600], ...
    'baseline', [-200 0], 'select_ch', 1:32);

for nsb=1:55
    fname = sprintf('./data/s%02d.mat', nsb);
    EEG = load(fname);
    eeg_train = EEG.train;
    
    % ------------------------ CALIBRATION EEG ------------------------- %
    feat_train = [];
    label_train = [];
    for nRun = 1:length(eeg_train)
        cur_eeg = eeg_train{nRun};
        data = cur_eeg.data(ch,:);
        cur_eeg.data = data;
        
        [feature_set, label_set] = preproc_P3speller(cur_eeg, Params_P3speller, 'labeled');
        feat_train = cat(1, feat_train, feature_set);
        label_train = cat(1, label_train, label_set);
    end
    
    % train SWLDA
    rd = shuffle([feat_train label_train]);
    feat_train = rd(:, 1:end-1);
    label_train = rd(:, end);
    mdl_SWLDA = trainSWLDA(feat_train, label_train, 0.08, 60);
    
    predictor = predictSWLDA(mdl_SWLDA, feat_train, 'label');
    acc_binary_epoch = sum(predictor==label_train) / length(label_train);
    fprintf('P300 speller train epoch classification accuracy is %.2f%%\n', 100*acc_binary_epoch);
    
    % ----------------------------- TEST EEG --------------------------- %
    eeg_test = EEG.test;
    % load classifier weight and display results
    for nRun = 1:length(eeg_test)
        cur_eeg = eeg_test{nRun};
        word_len = cur_eeg.nbTrials / (length(Configs_P3speller.seq_code)*Configs_P3speller.full_repeat);
  
        data = cur_eeg.data(ch,:); % remove extra channels for analysis
        cur_eeg.data = data;
        [feat_test, label_train] = preproc_P3speller(cur_eeg, Params_P3speller, 'labeled');
        
        rd = shuffle([feat_test label_train]);
        feat_test = rd(:, 1:end-1);
        label_train = rd(:, end);
        
        predictor = predictSWLDA(mdl_SWLDA, feat_test, 'label');
        acc_binary_epoch = sum(predictor==label_train) / length(label_train);
        %         fprintf('P300 speller epoch (binary) classification accuracy is %.2f%%\n', 100*acc_swlda);
        
        % ------------ letter detection accuracy
        markers_seq = cur_eeg.markers_seq;
        letter_idx = find(ismember(cur_eeg.markers_seq, Configs_P3speller.seq_code)); % seq len x # of repetition x # of words
        feat_unknown = preproc_P3speller(cur_eeg, Params_P3speller, 'unknown');
        
%         own classifier - predicted label for feat_unknwon
%         (predicted_labels) target: positive, non-target: negative
%         ans_letters_custom = detect_letter_P3speller_custom(predicted_labels, world_len, ...
%             cur_eeg.text_to_spell, letter_idx, markers_seq, Configs_P3speller);       
        ans_letters = detect_letter_P3speller(mdl_SWLDA, word_len, feat_unknown, ...
            cur_eeg.text_to_spell, letter_idx, markers_seq, Configs_P3speller);
        
        acc_on_repetition(nsb, nRun, :) = ans_letters.acc_on_repetition; % [nbsub, nbrun, nbrepetition(1~15)]
        correct_on_repetition(nsb, nRun, :) = ans_letters.correct_on_repetition; % [nbsub, nbrun, nbrepetition(1~15)]

        fprintf('User answer: %s (%d/%d), accuracy: %.2f\n', ...
            ans_letters.text_result, correct_on_repetition(nsb, nRun, 15), word_len, acc_on_repetition(nsb, nRun, 15));
    end
    fprintf('\nOverall: %d/%d, accuracy: %.2f\n\n', sum(correct_on_repetition(nsb, :, Configs_P3speller.full_repeat), 2), ...
        word_len*nRun, sum(correct_on_repetition(nsb, :, Configs_P3speller.full_repeat), 2)/(word_len*nRun));
    
end


save('Total_performance.mat', 'acc_on_repetition', 'correct_on_repetition');
%%

load('Total_performance.mat');

disp(size(correct_on_repetition));
total_corrected_repeat = squeeze(sum(correct_on_repetition, 2));
total_acc_repeat = total_corrected_repeat / 28;

avgd_acc = mean(total_acc_repeat, 1);

figure, 
h=boxchart(total_acc_repeat);
pbaspect([2 1 1]);
set(gca, 'fontsize', 13);
xlabel('The nubmer of repetitions');
ylabel('Letter detection accuracy');
hold on;
box on;
plot(avgd_acc, '-o');

% ylim([0 1.2]);
% % text(0.5, 1.1, num2str(round(avgd_acc(1), 2)));


%% bar plot
sky_boxchart = [0 0.4470 0.7410];
figure,
bar(total_acc_repeat(:, end), ...
    'FaceColor', sky_boxchart, 'FaceAlpha', 0.2, ...
    'EdgeColor', sky_boxchart, 'Linewidth', 1.2); 
set(gca, 'fontsize', 13);
xlabel('Subjects');
ylabel('Offline letter detection accuracy');
ylim([0.4 1]);
xlim([0 56]);
pbaspect([2 1 1]);

File: Python/P300speller_predict_letter.py
import mat73
import matplotlib.pyplot as plt
import numpy as np
from sklearn import linear_model
import statsmodels.api as sm

from functions.func_filters import butter_bandpass_filter
from sklearn.linear_model import LinearRegression
from functions import func_preproc as preproc
from functions import func_classifier as classifier

# pre-defined parameters
baseline = [-200, 0] # in ms
frame = [0, 600] # in ms

# One need to specify data directory
data_dir = "/Volumes/T5_2TB/Matlab_workspace/P3BCI2017_current/Won2021/data/"

nsub = 10
EEG = mat73.loadmat(data_dir+'s{:02d}.mat'.format(int(nsub)))

# PART 1 -  Pre-processing for training EEG
for n_calib in range(len(EEG['train'])):
    cur_eeg = EEG['train'][n_calib]
    data = np.asarray(cur_eeg['data'])
    srate = cur_eeg['srate']
    data = butter_bandpass_filter(data, 0.5, 10, srate, 4)
    markers = cur_eeg['markers_target']

    targetID = np.where(markers==1)[0]
    nontargetID = np.where(markers==2)[0]

    tmp_targetEEG = preproc.extractEpoch3D(data, targetID, srate, baseline, frame, False)
    tmp_nontargetEEG = preproc.extractEpoch3D(data, nontargetID, srate, baseline, frame, False)
    if n_calib == 0:
        targetEEG = tmp_targetEEG
        nontargetEEG = tmp_nontargetEEG
    else:
        targetEEG = np.dstack((targetEEG, tmp_targetEEG))
        nontargetEEG = np.dstack((nontargetEEG, tmp_nontargetEEG))

# data dimension for
print('Target EEG shape', targetEEG.shape) # ch x time x trial
print('Nontarget EEG shape', nontargetEEG.shape) # ch x time x trial

# training target data
down_target = preproc.decimation_by_avg(targetEEG, 24)
down_nontarget = preproc.decimation_by_avg(nontargetEEG, 24)

ch_target, frame_target, trial_target = down_target.shape
ch_nontarget, frame_nontarget, trial_nontarget = down_nontarget.shape

# ch x time x trial -> (ch* time) x trial -> trial x (ch*time)
feat_target = np.reshape(down_target, (ch_target*frame_target, trial_target)).transpose()
feat_nontarget = np.reshape(down_nontarget, (ch_nontarget*frame_nontarget, trial_nontarget)).transpose()

# labels - (+1) for target and (-1) for nontarget
y_target = np.ones((feat_target.shape[0],1))
y_nontarget = -np.ones((feat_nontarget.shape[0],1))

print('Target feature vector shape', feat_target.shape)
print('Nontarget feature vector shape', feat_nontarget.shape)

feat_train = np.vstack((feat_target, feat_nontarget))
y_train = np.vstack((y_target, y_nontarget))

# shuffle target and nontarget indices
np.random.seed(101)
idx_train = np.arange(feat_train.shape[0])
np.random.shuffle(idx_train)
feat_train = feat_train[idx_train, :]
y_train = y_train[idx_train, :]

# Train linear classifier and save the model - can apply own classification model
feat_column = np.array(range(feat_train.shape[1]))
feat_best_column, stats_best = classifier.stepwise_linear_model(feat_train, feat_column, y_train, 0.08)
print(feat_best_column.shape)

argsort_pval = np.argsort(stats_best.pvalues)
feat_selec_column = feat_best_column[argsort_pval[range(60)]] # will only use the best 60 features
feat_train_select = feat_train[:, feat_selec_column]

mdl_linear = LinearRegression()
mdl_linear.fit(feat_train_select, y_train)
# now mdl_linear is a model for the current subject (subject-specific)

pred_train = np.sign(mdl_linear.predict(feat_train_select))
print('Training accuracy (binary - target or nontarget): {:02f}'.format(np.sum(pred_train==y_train) / len(pred_train)))

# PART 2 - Calculate Letter detection accuracy with test EEG
# Ingredients for P300 speller
spellermatrix = ['A', 'B', 'C', 'D', 'E', 'F', 
                 'G', 'H', 'I', 'J', 'K', 'L',
                 'M', 'N', 'O', 'P', 'Q', 'R',
                 'S', 'T', 'U', 'V', 'W', 'X',
                 'Y', 'Z', '1', '2', '3', '4',
                 '5', '6', '7', '8', '9', '_']
Config_P3speller = {"seq_code": range(1,13), "full_repeat": 15, "spellermatrix": spellermatrix}
Params_P3speller = {"freq": [0.5, 10], "frame": [0, 600], "baseline": [-200, 0], "select_ch": range(1, 33)}

for n_test in range(len(EEG['test'])):
    cur_eeg = EEG['test'][n_test]
    data = np.asarray(cur_eeg['data'])

    srate = cur_eeg['srate']
    data = butter_bandpass_filter(data, 0.5, 10, srate, 4)
    markers = cur_eeg['markers_target']
    word_len = int(cur_eeg['nbTrials'] / (len(Config_P3speller['seq_code'])*Config_P3speller['full_repeat']))

    markers_seq = cur_eeg['markers_seq']
    letter_idx = np.where(np.isin(markers_seq, Config_P3speller['seq_code']))[0]

    unknownEEG = preproc.extractEpoch3D(data, letter_idx, srate, baseline, frame, False)
    down_unknown = preproc.decimation_by_avg(unknownEEG, 24)
    ch_unknown, frame_unknown, trial_unnknown = down_unknown.shape
    feat_unknown = np.reshape(down_unknown, (ch_unknown*frame_unknown, trial_unnknown))
    feat_unknown = feat_unknown.transpose()

    # opt - calculate the all classification results from feat_unknown
    pred_unknown = mdl_linear.predict(feat_unknown[:, feat_selec_column]) # your answers // can apply own classification algorithm
    ans_letters = preproc.detect_letter_P3speller(pred_unknown, word_len, cur_eeg['text_to_spell'], letter_idx, markers_seq, Config_P3speller)
    cur_text_result = ans_letters['text_result']

    print(f"User answer: {cur_text_result} ({int(ans_letters['correct_on_repetition'][-1])}/{int(word_len)}), accuracy: {ans_letters['acc_on_repetition'][-1]}")